File: src/main/java/dev/lain/claudejb/mcp/ToonCursor.java
package dev.lain.claudejb.mcp;

import dev.lain.claudejb.mcp.ToonLines.Line;
import java.util.ArrayList;
import java.util.List;

final class ToonCursor {

    private final List<Line> lines;
    private final boolean strict;
    private final List<Integer> spans = new ArrayList<>();
    private int index;

    ToonCursor(List<Line> lines, boolean strict) {
        this.lines = lines;
        this.strict = strict;
    }

    Line peek() {
        return index < lines.size() ? lines.get(index) : null;
    }

    Line next() {
        return lines.get(index++);
    }

    void skipBlanks() {
        boolean skipped = false;
        while (peek() != null && peek().blank()) {
            index++;
            skipped = true;
        }
        Line next = peek();
        if (next == null || spans.isEmpty()) return;
        if (skipped && strict && next.depth > spans.get(0)) {
            throw new ToonException("blank line inside the scope that continues at line " + next.number);
        }
    }

    void enterSpan(int depth) {
        spans.add(depth);
    }

    void leaveSpan() {
        spans.remove(spans.size() - 1);
    }
}

File: src/main/kotlin/dev/lain/claudejb/model/mcp/toon/ToonCursor.kt
package dev.lain.claudejb.model.mcp.toon

internal class ToonCursor(private val lines: List<Line>, private val strict: Boolean) {

    private var index = 0
    private val spans = ArrayList<Int>()

    fun peek(): Line? = lines.getOrNull(index)

    fun next(): Line = lines[index++]

    fun skipBlanks() {
        var skipped = false
        while (peek()?.blank == true) {
            index++
            skipped = true
        }
        val next = peek() ?: return
        val outer = spans.firstOrNull() ?: return
        if (skipped && strict && next.depth > outer) toonError("blank line inside the scope that continues at line ${next.number}")
    }

    fun enterSpan(depth: Int) {
        spans += depth
    }

    fun leaveSpan() {
        spans.removeAt(spans.lastIndex)
    }
}
